Add "newcontainer" hyper command to "start" runtime subcommand

jcvenegas opened the pull request:
This PR adds newcontainer hyper request   in start command. It still need to tests using a real shim and using docker.

After send newcontainer hyper request, the "start" flow continues as usual.

Commit:
proxy: Add the start of cc-proxy

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add systemd unit files

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add a hyper command

This command allows clients to execute hyperstart commands through the
proxy.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add a bit of documentation

Document the protocol so far as well as the systemd integration.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add a bye command

We need a way to for clients to tell the proxy they are finished with
the qemu/hyperstart instance, either by detecting qemu has closed the
sockets or by a more active way (eg. a bye command or keep alive
packets).

It's not that easy to detect that the a has closed the socket when using
the high level Go net functions. I'd use epoll() and EPOLLRDHUP but
there isn't really a way to plug it in.

So instead, let's try the `bye` command approach, at least for now.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
Initial rudimentary runtime proxy code.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Support a non socket-activated fallback

When wanting to quickly test something it's easier to just launch the
proxy rather than restarting the systemd controlled daemon.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
vendor: Vendor go-difflib

This library is needed of testify/assert

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
vendor: Vendor go-spew

This library is needed for testify/assert

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
gitignore: Ignore config.* only at the root for the directory

This was matching config.go, we really only want to ignore the autoconf
generated config.*.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
vendor: Add testify

The assert facilities in the go standard library are a bit underwhelming
is certainly too verbose compared to what other languages/frameworks
offer.

Testify and its assert package introduces a nice API to do asserts in
test.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
hyperstart: fix crash when cc-proxy is not running

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Move call to cc_proxy_wait_until_ready() to child.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
hyperstart: alloc and release proxy into state struct

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
proxy: Add tests for protocol.go

This tests the protocol itself as well as the API to define payloads on
top of the protocol.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
state: fix state vm pid

fix error: "failed to convert 'pid' to int"

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Proxy: Update for new message format and flush channel.

No main loop is used when talking to the proxy, so an explicit
flush is required to actually send the data.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Proxy: Move call to cc_proxy_wait_until_ready() to child.

cc_proxy_wait_until_ready() must be called by the child since it has
full hypervisor config information.

However, the logic is currently incorrect since the proxy comms need to
happen *after* the hypervisor has been launched (to ensure the
hyperstart sockets have been created).

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Use a struct to encapsulate the return state of a Handler

At the moment protocol handlers can return an error and a map that will
serialized to JSON in the answer.

I now would like to add the possibility of returning a file descriptor
as part of a protocol Response.

For that to be cleanly done, what the handler can return needs to be a
bit more than just a list of return values and it's time to introduce a
struct to store the states handlers can return.

This changes the handler API from:

func echoHandler(data []byte, userData interface{}) (map[string]interface{}, error) {
	/* do things *.
       return map[string]interface{}{
               "key": value,
       }, nil
}

To:

func echoHandler(data []byte, userData interface{}, response *HandlerResponse) {
	/* do things *.
	response.AddResult("result", echo.Arg)
}

This way, we can keep adding more things to the response struct without being
too invasive.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
WIP: hyperstart: modify create process

TODO: ...

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Process: Fix off-by-one bug causing VM to fail.

Buffer was too small resulting in random bytes as the end. This caused
g_strsplit() to fail to parse the last argument resulting in an invalid
hypervisor command-line.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Process: Simplify calculation of hypervisor args len.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
g_autofree() ftw!

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Add helpers for fd passing through an AF_UNIX socket

I'd like to give a fd to the client as part of the "allocateIO"
response.

Start by adding the necessary facilities to pass fds around.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
Proxy: Force write to proxy before read occurs.

Also, pass data to proxy message handlers.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Process: Remove hook_status_pipe.

The hypervisor child no longer needs to block on the hook status, so
remove the pipe.

Also, create the state file in the parent after the hypervisor and shim
are running.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
State: Fix hypervisor pid in state file.

Removed duplicate hypervisor pid variable.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Extract Socketpair() out of the testing code

Turns out I want to create a pair of already connected sockets to hand
one over to clients. Promote Socketpair out of the testing code.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
State: Create proxy object dynamically.

This allows cc_oci_config_update() to transfer the proxy state
correctly, fixing the incorrect proxy details in the state file.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Logging: Make socket connection errors meaningful.

Display the name/path for all sockets that cannot be connected to,
to make log analysis easier.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
removed old comment.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Config: Create main config object dynamically.

Creating the cc_oci_config object dynamically allows the new
constructor to also create the embedded proxy object. This is important
to allow the proxy object to be transferred from the state object to
the config object for non-"create" calls.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Tests: Update for dynamically-allocated config object.

All tests updated to reflect new design (except for test/oci_test.c).

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
wait for CTL socket creation

Unfortunately launching the hypervisor does not guarantee that
CTL and TTY exist, for this reason we MUST wait for them before
writing down any message into proxy's socket

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Add function header for cc_oci_config_create().

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix state file creation handling for config->proxy.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Update for handling dynamically-created config object.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
test_helper_create_state_file(): Update for handling config->proxy.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Update test for handling dynamically-created config object.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix test_cc_oci_state_file_create() which now sets config->proxy.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Updated state_test.c for config->proxy.

Updated existing json test files and created new ones for config->proxy.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fixed spec-handler tests to work with dynamically-created config object.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
cc_oci_start(): Remove now-unneeded SIGCONT.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Add an allocateIO payload and handle I/O streams

allocateIO requests the proxy to allocate seq numbers to use to identify
I/O streams between hyperstart and clients.

When a successful allocateIO returns, it is followed by a file
descriptor clients should use to receive and forward I/O to the proxy.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add an attach command

Clients need to be able to attach a new connection to a running VM
(anything but $runtime create will need this). The attach payload can be
used to do just that.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
Only set proxy variables if not already set.

This change helps keep the test code simple.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix valgrind errors and update to include explicit test on proxy values.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix valgrind-detected test memory leak.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
remove bad event message

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
fix 'start' hang up

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Remove unused "__unused__" macro.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Proxy: Added cc_proxy_free() function.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Allocate config->proxy in cc_oci_expand_cmdline().

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Call cc_proxy_free().

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix config->proxy handling.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Valgrind fixes and formatting.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix config->proxy handling.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Comments.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
proxy: Move Request and Response to the api package

These structures define the JSON snippets transporting payloads, so they
are really part of the proxy API.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
Proxy: WIP code.

Changes:

- Added code to check proxy response.
- Added function to request the proxy run a generic hyper command.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
WIP: implement create pod

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
Stop compiler warning about unused parameters.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Comments.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Check function param.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Don't hard-code proxy command.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Add message if proxy wait fails.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Check function parameters.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
cc_proxy_run_cmd() now validates proxy response.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Removed incorrect message.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Call cc-shim with arguments.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
fix communication with cc-proxy

fix Hyper format

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
proxy: Split a function serving a single client

We'll need that for unit testing, mocking a single connection.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Change hyper command example

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
proxy: Omit the request data field if nil

When sending a proxy payload that doesn't take any data, we can omit
specifying data in the JSON serialization.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
latest

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
tests: Add a mocking library, starting by hyperstart

We want to build full test harnesses around the various components to
fully test them. Start that journey by mocking hyperstart.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add a client API

Clients talk to the proxy with an API made of JSON packets. While we do
expose and document that level of the API, we can also provide
convenience go wrappers to make the RPC calls to the proxy. It should
make writing go programs (and tests!) much easier.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
proxy: Add tests covering the various proxy payload

It's high time to cover the front facing payloads. This commit uses the
mocked hyperstart and client API to build a test harness around a proxy
object.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

Commit:
latest

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
latest

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

jcvenegas committed:
add newcontainer proxy/hyperstart command

Modify start command to create a new container and
initial workload based in oci config.json file.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
Dont generate containerexec, hyperstart will handle it

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

Commit:
fix file descriptors bug

Commit:
Logging: don't write to stderr on error.

Cannot write to stderr since the underlying fd may be closed or used for
other purposes.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Shim: Improve fd checks.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Proxy: Improve channel checks.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Fix child log setup handling.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Process: Use pipe2(2) to avoid having to call cc_oci_fd_set_cloexec().

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

Commit:
Remove cruft.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

devimc commented on src/proxy.c:
```diff
@@ -732,3 +804,132 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
+       json_object_set_object_member (newcontainer_payload, "process", process);
+
+       if (! cc_proxy_run_hyper_cmd (config, "newcontainer", newcontainer_payload)) {
+               g_critical("failed to run pod create");
```
"pod create" ?  or "newcontainer"

jcvenegas replied:
ups, copy paste :(

jcvenegas committed:
state: Add oci process node to state file

start command require to know the initial process workload.
This patch adds the process workload information into the state file.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

grahamwhaley commented:
Just a note/nitpick @jcvenegas - some of the commit messages could do with a 'subsystem: ' start to their subject lines - I know @jodh-intel already has a bunch of cleaning up to do in his tree, so let's see if we can help him by not adding more :-)

And, in one subject s/dont/don't/

amshinde commented on src/proxy.c:
```diff
@@ -732,3 +804,132 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
+       if (! cc_proxy_connect (config->proxy)) {
+               goto out;
+       }
+       if (! cc_proxy_attach (config->proxy, config->optarg_container_id)) {
```
You will need the attachIO proxy command instead for the create-start flow.

amshinde commented on src/proxy.c:
```diff
@@ -732,3 +804,132 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
+       /* FIXME how to handle IO streams id's  ?*/
+       json_object_set_int_member (process, "stdio", 1);
+       json_object_set_int_member (process, "stderr", 2);
+
```
Need to replace these with the seq numbers received from the attachIO command.

amshinde commented on src/proxy.c:
```diff
@@ -732,3 +804,132 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
+
+	// FIXME use config cwd
+       json_object_set_string_member (process, "workdir", "/");
+
```
I had pushed a patch to set the home dir to the user's home dir. You may want to use that. Thats how runc sets its home directory.

amshinde commented on src/proxy.c:
```diff
@@ -732,3 +804,132 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
+       json_object_set_string_member (newcontainer_payload, "image", "");
+
+       /* newcontainer.process */
+       json_object_set_boolean_member(process, "terminal", true);
```
The terminal value present in config.json should be used here.

jcvenegas commented:
Note: Adding process from config.json into the state file will would break the start flow when is needed to wait or not (detach/attach).

https://github.com/01org/cc-oci-runtime/pull/336 fix the problem
